src/util/psbt/mod.rs

```diff
@@ -210,6 +213,234 @@ impl PartiallySignedTransaction {
+
+        let is_wpkh = script.is_v0_p2wpkh();
+        let is_wsh = script.is_v0_p2wsh();
+
+        let is_nested_wpkh = script.is_p2sh()
+            && input.redeem_script.as_ref()
+            .map(|s| s.is_v0_p2wpkh()).unwrap_or(false);
+
+        let is_nested_wsh = script.is_p2sh()
+            && input.redeem_script.as_ref()
+            .map(|x| x.is_v0_p2wsh()).unwrap_or(false);
+
+        let is_segwit = is_wpkh || is_wsh || is_nested_wpkh || is_nested_wsh;
+
+        let sighash = if is_segwit {
+            if is_wpkh || is_nested_wpkh {
+                let script_code = if is_wpkh {
+                    Script::p2wpkh_script_code(&script).ok_or(SighashError::NotWpkh)?
+                } else {
+                    Script::p2wpkh_script_code(input.redeem_script.as_ref().expect("checked above"))
+                        .ok_or(SighashError::NotWpkh)?
+                };
+                cache.segwit_signature_hash(input_index, &script_code, utxo.value, hash_ty)?
+
+            } else {
+                let script_code = input.witness_script.as_ref().ok_or(SighashError::MissingWitnessScript)?;
+                cache.segwit_signature_hash(input_index, script_code, utxo.value, hash_ty)?
+            }
+        } else {
+            let script_code = if script.is_p2sh() {
+                input.redeem_script.as_ref().ok_or(SighashError::MissingRedeemScript)?
+            } else {
+                &script
+            };
+            cache.legacy_signature_hash(input_index, script_code, hash_ty.to_u32())?
+        };
+
+        Ok((Message::from_slice(&sighash).expect("sighashes are 32 bytes"), hash_ty.into()))
+    }
+
+    /// Returns the spending utxo for this PSBT's input at `input_index`.
+    fn spend_utxo(&self, input_index: usize) -> Result<&TxOut, SighashError> {
+        let input = &self.inputs[input_index];
+        let utxo = if let Some(witness_utxo) = &input.witness_utxo {
+            &witness_utxo
+        } else if let Some(non_witness_utxo) = &input.non_witness_utxo {
+            let vout = self.unsigned_tx.input[input_index].previous_output.vout;
+            &non_witness_utxo.output[vout as usize]
+        } else {
+             return Err(SighashError::MissingSpendUtxo);
+        };
+        Ok(utxo)
+    }
+
+    /// Checks `input_index` is within bounds for the PSBT `inputs` array and
+    /// for the PSBT `unsigned_tx` `input` array.
+    fn check_index_is_within_bounds(&self, input_index: usize) -> Result<(), SighashError> {
+        if input_index >= self.inputs.len() {
+            return Err(SighashError::IndexOutOfBounds(input_index, self.inputs.len()));
+        }
+
+        if input_index >= self.unsigned_tx.input.len() {
+            return Err(SighashError::IndexOutOfBounds(input_index, self.unsigned_tx.input.len()));
+        }
+
+        Ok(())
+    }
+
+    fn signing_algorithm(&self, input_index: usize) -> Result<Algo, SighashError> {
+        self.check_index_is_within_bounds(input_index)?;
+
```

Comment:
Since we check the bounds higher up in `sign()`, do we need to do it here too? Same with `sighash()`? I guess this is pretty defensive as it is now so no big deal :)

```suggestion
```

Reply:
Yes this is defensive, I personally prefer to see a length check anywhere I see an array access.